### ISSUE TITLE
Support for Shimadzu LCD mass spectrometry streams

### ISSUE THREAD

**Issue by silasmellor** [ethanbass/chromConverter#11]
Hi, is it possible to add a parser for Shimadzu .lcd files? I have attached a file example for reference.
Best, Silas
[Anthocyanin_2_MeOH001.zip](https://github.com/ethanbass/chromConverter/files/10372505/Anthocyanin_2_MeOH001.zip)

**Comment by ethanbass** [comment]
I will try to look into it, though I can't really make any promises or give you a timeline. It's a little bit tricky since the format is not documented at all, as far as I know. It would be helpful if you could provide a little more information about your file (e.g. what kind of instrument/software it was produced with and what the detectors are on the machine). Thank you for sharing the file! 
Ethan

**Comment by silasmellor** [comment]
Hey Ethan, that is all i can ask! The file comes from Shimadzu Lab Solutions v 5.71. The instrument is a shimadzu LC-20 with PDA iirc, but i will double check and let you know.

**Comment by ethanbass** [comment]
Thanks. Also, you may already know this (and I know it's not as convenient as reading the `.lcd` files directly), but I believe you can export the raw PDA data from LabSolutions by right clicking the sample name and selecting `File Conversion:Convert to ASCII`. These ascii txt files do have a parser in chromConverter already.

Also, if you can generate the ascii file for the `.lcd` file you sent it would be helpful to have as a point of comparison.

**Comment by silasmellor** [comment]
Hey Ethan, i was actually not aware of that, that sounds excellent! I will try it and upload the ascii file for you.
Thanks again,
Silas

**Comment by silasmellor** [comment]
Hey Ethan, here is the corresponding ASCII from LabSolutions.
[Anthocyanin_2_MeOH001.txt](https://github.com/ethanbass/chromConverter/files/10423968/Anthocyanin_2_MeOH001.txt)

**Comment by ethanbass** [comment]
Thanks for providing this file! I don't know if you tried to read it yet using chromConverter, but it actually throws an error because it has commas as the decimal separator (european style) rather than periods (american style). I will put up a patch shortly that can read the european-style files.

**Comment by ethanbass** [comment]
This should be fixed by [725ddff](https://github.com/ethanbass/chromConverter/commit/725ddfff8aca822ede798db435dfb9accc4fd1bc). You should now be able to read in these ascii files at least:
```
x<-read_chroms(path_to_file_or_files, format_in="shimadzu_dad")
```

**Comment by silasmellor** [comment]
Hey Ethan, yep i did try it with the ascii and got an error, but didnt have time to figure out what it was about. Well, you fixed it before i even got that far so thanks once again!

**Comment by silasmellor** [comment]
I get a different error now:

> Warning in read_chroms(paths = "Anthocyanin_2_MeOH001.txt", format_in = "other") :
>   Error in py_call_impl(callable, dots$args, dots$keywords) : 
>   TypeError: Index(...) must be called with a collection of some kind, '' was passed

I feel i should apologize on behalf of european PCs, for some reason comma is default decimal here even though it means we have to jump through all sorts of hoops with our data files...

**Comment by ethanbass** [comment]
This is the old error -- it should be fixed in the latest version. Did you refresh your R session after reinstalling the package? Usually the old package is still cached if you install a new version of a package and don't reset your R session. The CRAN version will not be updated for a while, so that one will not work.

I'd suggest:
1. install new version from github: `remotes::install_github("ethanbass/chromConverter"`
2. Reset R session (In R studio, you can do this by selecting `Session:Restart R`).
3. Reload the package
4. Try again: `read_chroms(`Anthocyanin_2_MeOH001.txt`, format_in="shimadzu_dad")`

**Comment by silasmellor** [comment]
Yep, you're right, i had restarted the session before reinstalling from github but didn't think to do so after... works now!

**Comment by ethanbass** [comment]
Excellent! I'm glad to hear it is working now

**Comment by captainsailboat** [comment]
Fully willing to do what I can to help with adding a .lcd parser, it would make life a lot easier! Let me know if there is anything I can do to help.

**Comment by ethanbass** [comment]
Hi @captainsailboat. Do you need a parser for the DAD format or mass spec data? I think there are similarities but the formats aren't exactly the same.  I have been working a little bit on the DAD file I got from @silasmellor and have managed to partially decode it but there are still some things I haven't been able to figure out. Do you have any experience decoding binary files? If so, I could try to share with you what I have so far and you could have a look for yourself. Otherwise, it would be helpful to have a few more example .lcd files (along with the corresponding ASCII files) to check the validity of what I've worked out so far.

**Comment by ethanbass** [comment]
I think I have this mostly figured out. I uploaded the parser to the Shimadzu LCD branch if ya'll want to try it out. I can't figure out where the retention times are stored yet in the file but the absorbance values are correct for the file I received from @silasmellor (https://github.com/ethanbass/chromConverter/tree/shimadzu_lcd). It is also quite slow, it takes over a minute on my computer to parse one file. You can access the parser (for now) using the `read_shimadzu_lcd` function.

**Comment by ethanbass** [comment]
I believe the final retention time should be stored somewhere in these two streams, but I can't figure out yet how it's encoded.

[3D_data_item.txt](https://github.com/ethanbass/chromConverter/files/12173393/3D_data_item.txt)
[2D_data_item_stream.txt](https://github.com/ethanbass/chromConverter/files/12173385/2D_data_item_stream.txt)

**Comment by captainsailboat** [comment]
To be completely honest with you I am struggling to figure out how to utilize this. I have installed R-studio and opened the R file but am not sure how to actually carry out the conversion.

**Comment by captainsailboat** [comment]
Hey again,

I am grateful for your help in trying to add support for these Shimadzu
filetypes. The native software that comes with the equipment allows
conversion to mzXML. I can provide you with .lcd and .mzXML files for
experimentation if needed. However, the conversion of 1 .lcd file to .mzXML
yields 2 separate mzXML files with extensions Ev1 and Ev3 which can make
analysis a bit frustrating when having to switch back and forth.
At the end of the day, I am capable and have access to workstations where I
can analyze the data with the native software, though it would be amazing
if I could carry out this same analysis from the comfort of my home. Once
again, I appreciate you for working on this as an extension of what I can
imagine is already a busy life for you.

Cheers.

**Comment by ethanbass** [comment]
That would be helpful if you could send some example files! You should be able to upload them directly here (you may need to change the extensions to something github likes like .txt) or you could email them to me. Could you also describe in a little more detail what kind of data is contained in the two mzXML files and what form you're trying to get the data into for further analysis?
Ethan

**Comment by ethanbass** [comment]
Replaced by #46. I can't implement reliable support for this file type without several more examples.

**Comment by ethanbass** [comment]
@captainsailboat I implemented support for a variety of shimadzu mass spectrometry formats in v0.10.0 if you're still interested.
